### DISCUSSION TITLE
One Bootloader type should be enough ! ?.

### DISCUSSION THREAD

**Discussion by sloped-soarer** [pascallanger/DIY-Multiprotocol-TX-Module#886] [Polls]
Hello.
Due to the DIY origins of Multiprotocol  ( the clue is in the name DIY-Multiprotocol-TX-Module ), we have used STK500 as the serial bootloader protocol. Those who made an Atmel 328 based module will no doubt remember Optiboot which uses STK500. However the project has moved on and I would guess that STM32 based modules are now dominant.

The STM32f103 devices have a serial bootloader in ROM, selected by some pin conditions at startup.
The STM32f103 bootloader is not a STK500 bootloader. It is detailed in ST Application note AN3155.

So with STM32 based modules we now have a situation where we might need to use the STM bootloader, or if the module firmware is being upgraded in a Transmitter capable of doing such a task we could be using the STK500 bootloader.

The STK500 bootloader has some deficiencies. e.g. No error checking of communications (No parity, No block checking.) Which is why a read back of programming is usually performed. The STM32 bootloader is better in this respect.

As the project develops further (hopefully) a new bootloader may be needed to support the microcontrollers that are used.

Whilst ST Microelectronics might be happy for us to use a reverse engineered version of their bootloader in our code whilst we are buying STM32 devices, this might not always be the case. I am not a legal professional.

So the question is "Should we continue to use the STK500 Bootloader ?"

**Comment by pascallanger** [comment]
The STK500 BL is extremely small which is why we kept it.
On STM32 when you flash the fw using USB, it uses the STM32 protocol. When you flash from the radio it uses STK500.

I'm working on a new module and you might help with the BL implementation if you want.

**Reply by sloped-soarer** [reply]
Thanks for the offer. I'm not really a programmer or a hardware designer, but i dabble in both.
There is a reverse engineered STM32 serial bootloader ... easy to find, but i do not wish to prejudice any design decisions.
Changing a bootloader is a design decision that needs careful consideration, particularly with the investments people have made in their transmitters and modules. However My respect for the Manufacturers that make money out of us when we buy modules is not the highest. I wonder how many times Manufacturers have approached the Multi Team ?.
Consideration needs be given to the TX software side as well.
If you were wondering why I started this discussion. Well it all stemmed from trying to use the largely forgotten OrangeRX module with EdgeTx. I think we can say the module is not supported by EdgeTX in terms of upgrading firmware via the TX. :)

Maybe You should design your own Multi bootloader and Licence it ... generate money for Multi !!!.
